### PR TITLE
Ability to optionally disable the musicbrainz metadata source

### LINKED ISSUE #400: Optionally disable the MusicBrainz data source
I'd like to perform an series of imports using beets but instead of using the musicbrainz autotagger I'd like to write my own matcher as a plugin.
There is a config option `import.autotag`, but this disables the tagger completely.

@sampsyo has suggested another config option to handle this, but it might be a little messy to implement, so I'm asking for some feedback.
My current idea would be to put a config option check in `beets/autotag/hooks.py`, in each of the methods that can be extended in the plugins, e.g. 
https://github.com/sampsyo/beets/blob/master/beets/autotag/hooks.py#L534
The alternative would be to move the check higher up to somewhere like https://github.com/sampsyo/beets/blob/master/beets/autotag/match.py#L409, but this might require some larger changes to the plugin system.

The other problem I can see is that there is a significant reliance on musicbrainz track ids, e.g. https://github.com/sampsyo/beets/blob/master/beets/autotag/match.py#L450. Any chance of working around these without a large rewrite?

### PR DESCRIPTION
**Pull request by snejus**
## Description

Fixes #400.  
Related to #2686

As I mentioned in https://github.com/beetbox/beets/issues/400#issuecomment-1067666254, this adds a patch that allows to disable the MusicBrainz metadata source during the import process.

I had a little go at making the `MusicBrainz Options` section in the docs a little bit more consistent, and copied over the entire default config section for clarity.

## To Do

- [x] Documentation. (If you've add a new command-line flag, for example, find the appropriate page under `docs/` to describe it.)
- [x] Changelog. (Add an entry to `docs/changelog.rst` near the top of the document.)
- [ ] Tests. (Encouraged but not strictly required.)

### CODE REVIEW

File: docs/reference/config.rst [hunk 1]
```diff
@@ -716,6 +716,39 @@ Default: ``{}`` (empty).
 MusicBrainz Options
 -------------------
 
+Default configuration::
```
Comment (sampsyo) [inline docs/reference/config.rst:719]: Just for my own sanity, could we keep this PR's docs changes focused on the new option? Some refactoring, reordering, and adding examples would be nice too, but it will be easier to think about in a separate PR. Thank you!

**Review by sampsyo — COMMENTED** [review]
Awesome; thanks for getting this started! I actually think that it might be a little cleaner to handle this disabling in the `beets.autotag.hooks` module, which is the "dispatcher" that aggregates results from multiple sources. For example, this is the line that asks MB for matches:
https://github.com/beetbox/beets/blob/3f6da68af2444b77807692023abfd607f0146750/beets/autotag/hooks.py#L615

Maybe it would make a little more sense to skip those calls there, rather than having them all return None?

File: docs/reference/config.rst (hunk 1 above)
Comment (snejus) [reply docs/reference/config.rst:719]: Of course, I'll move them out to another branch!

File: beets/autotag/mb.py [hunk 2]
```diff
@@ -482,6 +482,9 @@ def match_album(artist, album, tracks=None, extra_tags=None):
     The query consists of an artist name, an album name, and,
     optionally, a number of tracks on the album and any other extra tags.
     """
+    if not config["musicbrainz"]["enabled"].get(bool):
```
[the remark this answers is not in the document]
Comment (snejus) [reply beets/autotag/mb.py:485]: That's helpful, thanks!

**Comment by snejus** [comment]
I do agree, will add the logic to `hooks.py` in a second.

**Comment by snejus** [comment]
It did end up being cleaner once the error handling logic got abstracted away.

File: beets/autotag/hooks.py [hunk 3]
```diff
@@ -609,25 +619,17 @@ def album_candidates(items, artist, album, va_likely, extra_tags):
...
 
+    common_args = [album, len(items), extra_tags]
     # Base candidates if we have album and artist to match.
     if artist and album:
-        try:
-            yield from mb.match_album(artist, album, len(items),
-                                      extra_tags)
-        except mb.MusicBrainzAPIError as exc:
-            exc.log(log)
+        yield from handle_exc(mb.match_album, artist, *common_args)
```
Comment (wisp3rwind) [inline beets/autotag/hooks.py:625]: In my opinion, this pattern would be much more readable as

```python
    if artist and album and config["musicbrainz"]["enabled"]:
        yield from handle_exc(mb.match_album, artist, *common_args)
```

with the check for MB being enabled removed from `handle_exc`. This would separate exception handling and checking whether to query musicbrainz somewhat more cleanly.

If not going this way, renaming `handle_exc` -> `invoke_mb` or similar would also help.

Comment (snejus) [reply beets/autotag/hooks.py]: Good point, @wisp3rwind. I've just renamed the function to `invoke_mb`

File: beets/autotag/hooks.py [hunk 4]
```diff
@@ -598,6 +598,16 @@ def tracks_for_id(track_id):
             yield t
 
 
+def invoke_mb(call_func, *args):
+    if not config["musicbrainz"]["enabled"]:
+        return ()
```
Comment (sampsyo) [inline beets/autotag/hooks.py:603]: I do sort of like @wisp3rwind's suggestion to consider moving this out of the `invoke_mb` utility—it might make it a little clearer what's going on!

File: beets/autotag/hooks.py [hunk 5]
```diff
@@ -598,6 +598,16 @@ def tracks_for_id(track_id):
...
+
+    try:
+        return call_func(*args)
+    except mb.MusicBrainzAPIError as exc:
+        exc.log(log)
```
Comment (sampsyo) [inline beets/autotag/hooks.py:608]: I think this might cause a crash when an exception occurs. Namely, it seems like this function needs to return an iterable, right? So I think this exception handler should probably `return ()` or similar…

File: beets/autotag/hooks.py [hunk 6]
```diff
@@ -609,25 +619,17 @@ def album_candidates(items, artist, album, va_likely, extra_tags):
     constrain the search.
     """
 
+    common_args = [album, len(items), extra_tags]
```
Comment (sampsyo) [inline beets/autotag/hooks.py:622]: I know this might be bikeshedding, but I actually think the repetition (passing the same set of arguments to the two calls to `mb.match_album`) is a little more intelligible.

File: beets/autotag/hooks.py [hunk 7]
```diff
@@ -609,25 +619,17 @@ def album_candidates(items, artist, album, va_likely, extra_tags):
...
     # Base candidates if we have album and artist to match.
     if artist and album:
```
Comment (sampsyo) [inline beets/autotag/hooks.py:624]: This stanza and the next could share a single `if config["musicbrainz"]["enabled"]:`.

File: docs/reference/config.rst [hunk 8]
```diff
@@ -716,6 +716,17 @@ Default: ``{}`` (empty).
...
+.. _musicbrainz.enabled:
+
+enabled
+~~~~~~~
+
+This option allows you to disable using MusicBrainz as a metadata source. This applies
+if you use plugins that fetch data from alternative sources and should make the import
+process quicker.
+
+Default: ``yes``.
```
Comment (sampsyo) [inline docs/reference/config.rst:728]: I think this documentation stuff might be out of order—it makes the "top-level" description below appear to be part of the "enabled" subsection. I think this should appear among the other subsections below (which start with "searchlimit").

**Review by sampsyo — COMMENTED** [review]
Nice; looking good overall! I have a few low-level suggestions.

File: beets/autotag/hooks.py (hunk 4 above)
Comment (wisp3rwind) [reply beets/autotag/hooks.py:603]: I think I'm also fine with the current version (renamed to `invoke_mb`); I was mostly concerned about `handle_exc` being actively misleading since it does more than handling exceptions. So, from my side, I'd say it's up to you @snejus which variant you prefer.

File: beets/autotag/hooks.py (hunk 5 above)
Comment (wisp3rwind) [reply beets/autotag/hooks.py:608]: Nice catch! Yes, I'd suspect the same.

Comment (snejus) [reply beets/autotag/hooks.py:608]: Well caught! Fixed in e1ffadb2

File: docs/reference/config.rst (hunk 8 above)
Comment (snejus) [reply docs/reference/config.rst]: Agree - have moved it below.

**Review by sampsyo — APPROVED** [review]
Looks great to me; thank you for your attention to detail here! 😃
